src/schema/posts.ts

```diff
@@ -1371,6 +1391,58 @@ export const resolvers: IResolvers<any, Context> = {
 
       return { _: true };
     },
+    swapPinnedPosts: async (
+      _,
+      { id, swapWithId }: SwapPinnedPostArgs,
+      ctx: Context,
+    ): Promise<GQLEmptyResponse> => {
+      await ctx.con.transaction(async (manager) => {
+        const repo = manager.getRepository(Post);
+        const post = await repo.findOneBy({ id });
+        const swapWithPost = await repo.findOneBy({
+          id: swapWithId,
+        });
+
+        await ensureSourcePermissions(
+          ctx,
+          post.sourceId,
+          SourcePermissions.PostPin,
+        );
+
+        if (!post.pinnedAt || !swapWithPost.pinnedAt) {
+          throw new ValidationError('Posts must be pinned first');
+        }
+
+        const isNextPost = swapWithPost.pinnedAt > post.pinnedAt;
+        const swapPinnedTime = new Date(
+          swapWithPost.pinnedAt.getTime() + 1000 * (isNextPost ? 1 : -1),
+        );
+        const operation = isNextPost
+          ? `+ interval '1 second'`
+          : `- interval '1 second'`;
+        const comparison = isNextPost ? '>=' : '<=';
+
+        await manager.query(
+          `
+          UPDATE post
+          SET "pinnedAt" = "pinnedAt" ${operation}
+          WHERE "pinnedAt" IS NOT NULL
+            AND "pinnedAt" ${comparison} $1
+            AND "sourceId" = $2
+        `,
+          [swapPinnedTime, post.sourceId],
+        );
+
+        await repo.update(
+          { id },
+          {
+            pinnedAt: swapPinnedTime,
+          },
+        );
```

Comment:
Can probably run this and above query in transaction to perform only single write.

Reply:
are you referring to merging this update into the same query as above? That's a good point, but I think it would make the query a bit more complex and would not necessarily result in not having the equivalent of running 2 queries in performance.

unless you are referring to wrapping this in a transaction, in which case that is already in place 👍

Reply:
I didn't see it at the top, ignore me you did it perfect! 👌